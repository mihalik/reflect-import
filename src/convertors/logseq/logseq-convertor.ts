import {markdownToHtml} from 'helpers/markdown'
import {validateNotes} from 'helpers/validate'

import {tryParseDate} from './logseq-helpers'
import {LogseqBlock, LogseqConversionError, LogseqNote} from './types'
import {
  Backlink,
  ConvertedNote,
  ConvertOptions,
  Convertor,
  ConvertResponse,
  REFLECT_HOSTNAME,
} from '../../types'

export class LogseqConvertor implements Convertor {
  graphId: string
  linkHost: string
  noteIds: Record<string, string> = {}

  constructor({
    graphId,
    linkHost = REFLECT_HOSTNAME,
  }: {
    graphId: string
    linkHost?: string
  }) {
    this.graphId = graphId
    this.linkHost = linkHost
  }

  accept = {'application/json': ['.json']}

  convert({data}: ConvertOptions): ConvertResponse {
    const parsed = JSON.parse(data)

    if (parsed.version !== 1) {
      throw new LogseqConversionError('Only able to convert Logseq file version 1')
    }

    const notes = parsed.blocks as LogseqNote[]
    // Create a map of page names to ids.  We use this for the link resolver and adding
    // ids to backlinks.
    this.noteIds = notes.reduce(
      (acc, note) => ({...acc, [note['page-name']]: note.id}),
      {},
    )

    const convertedNotes = notes.map((note) => this.convertLogseqNote(note))

    return validateNotes(convertedNotes)
  }

  /**
   * Converts an individual logseq note into a ConvertedNote.  A logseq note has
   * the page name and the blocks (children).  We convert all the children to HTML
   * and then add the page name as an H1 to the top of the note.
   */
  private convertLogseqNote(note: LogseqNote): ConvertedNote {
    const {id, children} = note
    const subject = note['page-name']

    // Get all the HTML and backlinks for the current note
    const {html, backlinks} = this.parseBlocks(children)
    // Update the backlinks with the note ids from the page name to id map
    const updatedBacklinks = backlinks.map((b) => ({
      label: b.label,
      id: this.noteIds[b.label],
    }))

    return {
      id,
      html: `<h1>${subject}</h1>${html}`,
      subject,
      backlinks: updatedBacklinks,
      dailyAt: tryParseDate(subject),
    }
  }

  /**
   * Converts a list of logseq blocks into HTML.  This is a recursive through the
   * makeHtml function.  Each block is an li in a ul with the block's content in the
   * li.
   */
  private parseBlocks(blocks: LogseqBlock[]): {html: string; backlinks: Backlink[]} {
    const blockHtmls: string[] = []
    const backlinkCollection: Backlink[] = []

    // Get the HTML and backlinks for each block
    for (const block of blocks) {
      const {html, backlinks} = this.makeHtml(block)
      blockHtmls.push(`<li>${html}</li>`)
      backlinkCollection.push(...backlinks)
    }

    return {
      html: `<ul>${blockHtmls.join('')}</ul>`,
      backlinks: backlinkCollection,
    }
  }

  /**
   * Converts a single logseq block's content into HTML.  This will recursively
   * call parseBlocks if the block has children.
   */
  private makeHtml(block: LogseqBlock): {html: string; backlinks: Backlink[]} {
    // Logseq supports markdown and org.  I think doing the org conversion will be
    // more difficult.  So I'm going to start with markdown.
    if (block.format !== 'markdown') {
      throw new LogseqConversionError(
        'Logseq importer is only able to import markdown blocks',
      )
    }

    // Get the data for the current block
    const {html, backlinks} = markdownToHtml(block.content, {
      graphId: this.graphId,
      linkHost: this.linkHost,
      pageResolver: (pageName) => this.noteIds[pageName] ?? pageName,
    })

    // If the block has children then we need to get the html for the children
    let childHtml = ''
    let childBacklinks: Backlink[] = []
    if (block.children) {
      const childBlocks = this.parseBlocks(block.children)
      childHtml = childBlocks.html
      childBacklinks = childBlocks.backlinks
    }

    // We just append the child html to this current block HTML
    return {
      html: `${html}${childHtml}`,
      backlinks: [...backlinks, ...childBacklinks],
    }
  }
}
