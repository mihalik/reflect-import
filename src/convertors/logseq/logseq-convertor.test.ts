import {describe, it, expect} from 'vitest'

import exampleExport from './fixtures/logseq-example.json'
import {LogseqConvertor} from './logseq-convertor'

describe('LogseqConvertor', () => {
  it('parses example', () => {
    const convertor = new LogseqConvertor({graphId: '123'})
    const {notes} = convertor.convert({data: JSON.stringify(exampleExport)})

    expect(notes).toMatchSnapshot()
  })
})
