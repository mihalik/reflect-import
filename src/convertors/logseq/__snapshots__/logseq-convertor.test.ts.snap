// Vitest Snapshot v1

exports[`LogseqConvertor > parses example 1`] = `
[
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Contents</h1><ul><li><ul></ul></li></ul>",
    "id": "048b84ca-e374-40c9-90ba-1129207b1d88",
    "subject": "Contents",
  },
  {
    "backlinks": [
      {
        "id": "64530470-0fe2-4d95-915e-b67a1984c24c",
        "label": "May 2nd, 2023",
      },
      {
        "id": undefined,
        "label": "May 4th, 2023",
      },
    ],
    "dailyAt": 1683072000000,
    "html": "<h1>May 3rd, 2023</h1><ul><li><p>This is a test graph</p><ul></ul></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/64530470-0fe2-4d95-915e-b67a1984c24c\\">May 2nd, 2023</a></p><ul><li><p>This is some notes for yesterday</p><ul></ul></li></ul></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/May 4th, 2023\\">May 4th, 2023</a></p><ul><li><p>This is some notes for tomorrow</p><ul></ul></li></ul></li></ul>",
    "id": "64530465-ccd1-4b0e-8808-01b031fa10d8",
    "subject": "May 3rd, 2023",
  },
  {
    "backlinks": [],
    "dailyAt": 1682985600000,
    "html": "<h1>May 2nd, 2023</h1><ul><li><p>This is yesterday</p><ul></ul></li></ul>",
    "id": "64530470-0fe2-4d95-915e-b67a1984c24c",
    "subject": "May 2nd, 2023",
  },
  {
    "backlinks": [
      {
        "id": "645304ab-3a8e-41f6-996c-3e023a731b63",
        "label": "Another Page",
      },
    ],
    "dailyAt": undefined,
    "html": "<h1>New Page</h1><ul><li><p>This is a standalone page.</p><ul></ul></li><li><p><a class=\\"backlink new\\" href=\\"https://reflect.app/g/123/645304ab-3a8e-41f6-996c-3e023a731b63\\">Another Page</a></p><ul></ul></li><li><p>:properties:</p><ul></ul></li><li><ul></ul></li></ul>",
    "id": "6453049e-5817-4a48-8ad2-8a64c56295d6",
    "subject": "New Page",
  },
  {
    "backlinks": [],
    "dailyAt": undefined,
    "html": "<h1>Another Page</h1><ul><li><ul></ul></li><li><p>Deeply</p><ul><li><p>Nested</p><ul><li><p>Text</p><ul><li><p>Is</p><ul><li><p>Here</p><ul></ul></li></ul></li></ul></li></ul></li></ul></li><li><p><a href=\\"http://example.com\\">external link</a></p><ul></ul></li><li><p><img src=\\"../assets/SuccessKid_1683162562594_0.jpg\\" alt=\\"SuccessKid.jpg\\"></p><ul></ul></li><li><p>By Originally from the mother's Flickr page, found on Know Your Meme, Fair use, <a href=\\"https://en.wikipedia.org/w/index.php?curid=46523616\\">https://en.wikipedia.org/w/index.php?curid=46523616</a></p><ul></ul></li></ul>",
    "id": "645304ab-3a8e-41f6-996c-3e023a731b63",
    "subject": "Another Page",
  },
]
`;
